### PR TITLE
Int specs for views and fix n+1 problems

### PR DESCRIPTION
**Pull request by Apatakiti**
## Project requirements

**n+1**
Make sure that the N+1 problem is solved when fetching all posts and their comments for a user.
Do not use the bullet gem in this exercise - instead analyze the logs as showed in this [article](https://dev.to/junko911/rails-n-1-queries-and-eager-loading-10eh)
In your PR description include the screenshots of your app's logs before and after the fix
This is the corresponding wireframe view: blog all user posts

**Integration specs**

Use Capybara to write integration tests for each view in your project.
Think on the flow of things and group the tests of page in a way that makes sense. For example, one test would check for the things in the page and another test for clicking links on it.

**User index page:**

I can see the username of all other users.
I can see the profile picture for each user.
I can see the number of posts each user has written.
When I click on a user, I am redirected to that user's show page.

**user show page:**

I can see the user's profile picture.
I can see the user's username.
I can see the number of posts the user has written.
I can see the user's bio.
I can see the user's first 3 posts.
I can see a button that lets me view all of a user's posts.
When I click a user's post, it redirects me to that post's show page.
When I click to see all posts, it redirects me to the user's post's index page.

**User post index page:**

I can see the user's profile picture.
I can see the user's username.
I can see the number of posts the user has written.
I can see a post's title.
I can see some of the post's body.
I can see the first comments on a post.
I can see how many comments a post has.
I can see how many likes a post has.
I can see a section for pagination if there are more posts than fit on the view.
When I click on a post, it redirects me to that post's show page.

**Post show page:**

I can see the post's title.
I can see who wrote the post.
I can see how many comments it has.
I can see how many likes it has.
I can see the post body.
I can see the username of each commentor.
I can see the comment each commentor left.

Be sure to include integration specs for any other features you may have implemented!

Screenshot (123)

### CODE REVIEW

File: spec/views/user_index_spec.rb [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+
+  it 'should show the user name' do
+    expect(page).to have_content(@user.name)
+  end
+
+  it 'shows number of comments a post has' do
+    expect(page.body).to include('Comments: 3')
+  end
+
+  describe 'a user show page' do
```
Comment (V-Blaze) [inline spec/views/user_index_spec.rb:45]: - You have done great so far. :+1: :+1: But there are still requirements you need to meet before moving to the next milestone. :+1: :waffle: 

- [x] This is meant to be the test for `User index page:` but it seems you are having the same test you wrote for the `User show page:` here. This is not correct, I'll suggest you take a look at the requirement for the user index tests and follow that. Looking at the requirement, you are meant to be testing the user's index here which is the page that lists all the users including their `names, profile pictures, No. of posts`. I'll suggest your team works on this correctly. 

- - Index page requirement:
Selection_070

- - Remember, you are testing to make sure that all the user's data are displaying, so its meant to have multiple users and the assertion should be checking for eg: name accross all the users.
s

File: spec/views/user_show_spec.rb [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+    end
+    it 'shows part of a post body (This is my first post)' do
+      expect(page).to have_content('This is my first post')
+    end
+    it 'Not to show the Pagination button' do
+      expect(page.body).not_to include('Pagination')
+    end
+    it 'Redirect to post show page when a post is clicked' do
+      click_link(@last_post.title)
+      expect(current_path).to eq(user_post_path(@user, @last_post))
```
Comment (V-Blaze) [inline spec/views/user_show_spec.rb:42]: - [ ] You have done well writing the test for the user's show, but you re missing some of the required tests which  are important, I have pointed them out in the screenshot below, I'll suggest you look at the requirements and implement these. :+1: :+1: 
- - User's show requirements:
Selection_071

- [x] Also these tests seems not to be in the requirements for the `users show page` I suggest you strictly try to follow the requirement. 
```ruby
it 'shows number of comments a post has' do
      expect(page.body).to include('Comments: 3')
    end
    it 'shows number of likes a post has' do
      expect(page.body).to include('Likes: 0')
    end
    it 'shows the title of the post' do
      expect(page).to have_content(@last_post.title)
    end
    it 'shows part of a post body (This is my first post)' do
      expect(page).to have_content('This is my first post')
    end
```

File: spec/views/post_index_spec.rb [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+    expect(page).to have_content(@any_post.title)
+  end
+
+  it 'shows part of a post body (This is any post)' do
+    expect(page).to have_content('This is any post')
+  end
+
+  it 'shows the user name of a comment' do
+    expect(page).to have_content('Davon')
+  end
```
Comment (V-Blaze) [inline spec/views/post_index_spec.rb:39]: - [x] You have done well writing the test for the `User` **post index page**:`, but you're missing some of the required tests which  are important, I have pointed them out in the screenshot below, I'll suggest you look at the requirements and implement these. :+1: :+1: 

- - Post index requirements:
Selection_072

- [x] Also your describe has `describe 'posts/show.html.erb', ` but this is meant to be the post index test. I'll suggest you always make sure you are writing tests for the correct views. :wave: :small_airplane:

File: spec/views/post_show_spec.rb [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+      expect(page).to have_content('This is my first post')
+    end
+    it 'Not to show the Pagination button' do
+      expect(page.body).not_to include('Pagination')
+    end
+    it 'Redirect to post show page when a post is clicked' do
+      click_link(@last_post.title)
+      expect(current_path).to eq(user_post_path(@user, @last_post))
+    end
+  end
```
Comment (V-Blaze) [inline spec/views/post_show_spec.rb:44]: - Just as i have pointed out above, some of the required tests are missing here too. I'll suggest you carefully go through the requirements and include all. :+1: :wave: 
- - Post show requirements:
Selection_073

- [x] this test `    it 'shows part of a post body (This is my first post)' do
      expect(page).to have_content('This is my first post')
    end` should be checking that the post body is visible and not just part of the body. I'll suggest you update this. :+1: :100: 

- [x] Also this is already the post show page, this test `'Redirect to post show page when a post is clicked'` is redundant here. :+1: :100:

File: app/controllers/posts_controller.rb [hunk 5]
```diff
@@ -1,9 +1,30 @@
 class PostsController < ApplicationController
   def index
-    @posts = Post.all
+    @user = User.find(params[:user_id])
   end
```
Comment (V-Blaze) [inline app/controllers/posts_controller.rb:4]: - 
Selection_074

- [x] According to the project requirement, you are meant to fix the n+1 problem. The problem usually occurs when getting a users posts and subsequently fetching the comments. You are meant to update your `post index` method to fix that, then take a screenshot of the log when you visit /users/1/posts and add it to this pull request summary. I'll suggest you do that. :+1: :wave:

**Review by V-Blaze — CHANGES_REQUESTED** [review]
## Hi @Ademola-coding, @Rafiduzzaman, @mhnazary 

## Required changes 🖊️ 
# Almost there 🟡 🟦 🟨 

giphy


Great work so far, You have come a long way, keep moving forward👏 👍  
You've done well implementing the project requirements, but there are still some issues I think you need to improve on. 😄 👋 


 ## Highlights
- Added ``capybara` gem✔
-  Added some required specs👋 
-  Correct git flow✔

 ## Required Changes ♻️

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: spec/views/user_index_spec.rb (hunk 1 above)
Comment (Apatakiti) [reply spec/views/user_index_spec.rb:45]: Screenshot (130)

made proper correction

File: spec/views/user_show_spec.rb (hunk 2 above)
Comment (Apatakiti) [reply spec/views/user_show_spec.rb:42]: Screenshot (131)

made proper correction

File: spec/views/post_index_spec.rb (hunk 3 above)
Comment (Apatakiti) [reply spec/views/post_index_spec.rb:39]: Screenshot (133)

made proper corrections

File: spec/views/post_show_spec.rb (hunk 4 above)
Comment (Apatakiti) [reply spec/views/post_show_spec.rb:44]: Screenshot (134)

made proper corrections

**Pull request merged**